### ISSUE TITLE
Cached InlineCalculations contain both outputs when the code changed

### ISSUE THREAD

**Issue by greschd** [aiidateam/aiida_core#1871]
When enabling caching for an InlineCalculation and then changing the code of that calculation, the resulting calculation will have both the previous and the current output.

Example:

``` python
@make_inline
def test_inline():
    return Int(1)

test_inline()

@make_inline
def test_inline():
    return Int(2)

_, res = test_inline # res == {u'result': 2, u'result_<pk>': 1}
```

This seems to imply that InlineCalculations are not properly cached, i.e. they are still executed in some way.

**Comment by sphuber** [comment]
It is not so much a derivative of the workfunction, as it's functionality is just wrapped again by the same decorator logic, such that it also becomes a `FunctionProcess` and get's all the same attributes related to it being a `Process` sub class. I think the change that broke this is that I changed the way the source code of a `FunctionCalculation` (that includes `InlineCalculation` nodes) is stored. The source code used to be stored as an attribute. However, since this was deemed to database heavy, and the source code does not really need to be queryable, we opted to store the source code in the repository in stead. I guess that removing the source code from the attributes, now makes the hash independent of the source code, which might explain the behavior that you are seeing. Do you think that if we reintroduce the source code as a variable for the hash of a `FunctionCalculation` the caching might work again?

**Comment by greschd** [comment]
I think there are two separate issues here:

* The hashing doesn't seem to include the code. This seems odd to me, because in principle the content of the repository should also be part of the hash.
* Even when a cache match is (correctly) found, the inline calc still seems to execute, and add the "new" output. To see this, consider the test I wrote here: https://github.com/greschd/aiida_core/commit/180263f8bcdfdfd981fd46932757b18f7efff518 Even though the calculation is (correctly, since I did a sneaky update) loaded from the cache, it still gets the new output.

**Comment by greschd** [comment]
The second issue could probably be resolved by checking if the calculation is already finished at the correct spot in the ``FunctionCalculation``, similar to the way it's done for the ``JobCalculation``.

**Comment by giovannipizzi** [comment]
With the new provenance_redesign changes, probably  CalculationFunction(s) (that used to be InlineCalculations) can/should be cached. WorkFunctions should not, instead. 
However, it would be good to test that the example provided above now works with the new redesign (or to fix it).

**Comment by sphuber** [comment]
I just tested this on the current `provenance_design` and it is fixed:
```
from aiida.orm.data.int import Int
from aiida.work import calcfunction

@calcfunction
def test_inline():
    return Int(1)

test_inline()

@calcfunction
def test_inline():
    return Int(2)

res, node = test_inline.run_get_node()

node.get_outputs_dict()
> {u'result': 2, u'result_7': 2}
```

The workfunctions do not work correctly yet, but I am working on this in #2171

**Comment by giovannipizzi** [comment]
Great. Only question then is if we should merge @greschd 's PR in 0.12.x? Or this would make the behaviour even more inconsistent? (<=0.12.2 cache (wrongly), 0.12.3 does not cache, 1.0 caches with different names)

**Comment by sphuber** [comment]
Now also "fixed" (as in disabled) in `v0.12.3` in PR #1872
